### ISSUE TITLE
Timeout error on query if timeout=1 in a distributed deployment

### ISSUE THREAD

**Issue by mautini** [qdrant/qdrant#5463]
## Current Behavior
When querying the Qdrant Recommendation API, if the collection is distributed (across multiple shards) and the timeout is set to 1 second, Qdrant may return a timeout error during request distribution, even if the request execution time is less than 1 second. Without a timeout specified, or with timeout=2, the request succeeds.

## Steps to Reproduce
1.	Create a Qdrant cluster with multiple nodes using the official Helm Chart.
2.	Create a collection with multiple shards.
3.	Execute a simple recommend query like the one below (with timeout=1):
```
POST /collections/<collection_name>/points/query?timeout=1
{
  "query": {
    "recommend": {
      "positive": [
        "c026bde1-8399-488d-954a-c777a42ab74a"
      ]
    }
  },
  "with_payload": false,
  "with_vector": false
}
```
4. Qdrant output the following error
```
{
  "error": "Service internal error: 2 of 2 read operations failed:\n  Timeout error: Deadline Exceeded: status: DeadlineExceeded, message: \"Timeout: Timeout error: Operation 'Search' timed out after 0 seconds\", details: [], metadata: MetadataMap { headers: {\"content-type\": \"application/grpc\", \"date\": \"Mon, 18 Nov 2024 17:28:13 GMT\", \"content-length\": \"0\"} }\n  Timeout error: Deadline Exceeded: status: DeadlineExceeded, message: \"Timeout: Timeout error: Operation 'Search' timed out after 0 seconds\", details: [], metadata: MetadataMap { headers: {\"content-type\": \"application/grpc\", \"date\": \"Mon, 18 Nov 2024 17:28:13 GMT\", \"content-length\": \"0\"} }"
}
```

## Expected Behavior
The request should return a result without any errors. (Running the same request without a timeout set returns a response in less than 1 second.)

## Possible Solution
The issue seems to stem from the timeout settings between nodes (shards), as indicated by the message timed out after 0 seconds. This might be due to subtracting 1 (an integer) instead of a float (less than 1). The issue could be related to the following commits:
- https://github.com/qdrant/qdrant/pull/4844
- https://github.com/qdrant/qdrant/pull/4849 

## Context (Environment)
- **Deployment**: Qdrant deployed in a private Kubernetes cluster using the official Helm Chart.
- **Affected Versions**: At least Qdrant 1.11.5 and 1.12.3. It worked fine with 1.10.1
- **Reproducibility**: The issue is not related to any specific client, as it can be reproduced using Python or Curl.
- **Scope**: The issue occurs only with distributed clusters (multiple shards).

cc @evelynegroen

**Comment by mautini** [comment]
We are way below 1 second without the timeout parameter: 

```
{
  "result": {
    "points": [<REDACTED>]
  },
  "status": "ok",
  "time": 0.008113211
}
```

Also, we did not encounter this issue with version 1.10.x; it appeared after upgrading to 1.11.x

**Comment by generall** [comment]
if you have anything stored on disk (vectors / indexes / e.t.c.) first request might time out just because it is slower than 1 second cause it needs to fill the cache. Also, reported time doesn't include networking

**Comment by mautini** [comment]
Here is my collection config: 

```json
{
  "params": {
    "vectors": {
      "size": 384,
      "distance": "Cosine"
    },
    "shard_number": 3,
    "replication_factor": 2,
    "write_consistency_factor": 1,
    "on_disk_payload": true
  },
  "hnsw_config": {
    "m": 32,
    "ef_construct": 256,
    "full_scan_threshold": 10000,
    "max_indexing_threads": 0,
    "on_disk": false,
    "payload_m": 32
  },
  "optimizer_config": {
    "deleted_threshold": 0.2,
    "vacuum_min_vector_number": 1000,
    "default_segment_number": 0,
    "max_segment_size": null,
    "memmap_threshold": null,
    "indexing_threshold": 20000,
    "flush_interval_sec": 5,
    "max_optimization_threads": null
  },
  "wal_config": {
    "wal_capacity_mb": 32,
    "wal_segments_ahead": 0
  },
  "quantization_config": null,
  "strict_mode_config": {
    "enabled": false
  }
}
```

- I don't use any filter in my query
- I use `"with_payload": false,` so I assume payload on disk doesn't have any impact here.
- `it needs to fill the cache` --> If I rerun the query multiple time I still have the issue
- Chrome report 69ms (this include network time)
<img width="801" alt="image" src="https://github.com/user-attachments/assets/3d1e5db9-86a0-431c-b1a2-14e4e435bd58">

**Comment by generall** [comment]
ok, yeah, it seems I can reproduce it locally.

**Comment by agourlay** [comment]
@mautini Thanks a lot for the investigation!

I have some ideas how to fix this which I will share on the existing PR https://github.com/qdrant/qdrant/pull/5475

**Comment by zeyaddeeb** [comment]
can you guys please treat this as a patch for 1.12.4 or do a new release?

**Comment by timvisee** [comment]
> can you guys please treat this as a patch for 1.12.4 or do a new release?

We estimate to release a patch for this next week.

In the meanwhile you can do either of two things:
- use a higher timeout for the time being
- use the `qdrant/qdrant:dev` image

**Comment by zeyaddeeb** [comment]
I tried higher timeout with no success. I'll switch to `dev` image for now. Thank you!

**Comment by zeyaddeeb** [comment]
I was on `0.9` for a while and this hasn't been an issue until I upgraded

**Comment by zeyaddeeb** [comment]
I have to clarify.. qdrant is being called around `~800req/s`

**Comment by timvisee** [comment]
At that rate I don't expect this issue report to be related. Your searches are likely 'just' timing out because they are taking long. The upcoming patch release will not help with that.

Please have a look at the following documentation page if you're interested in tuning performance: https://qdrant.tech/documentation/guides/optimize/

**Comment by agourlay** [comment]
I agree that it is most likely unrelated.

This is why I recommend you to open a new issue with the details of your setup and query for us to investigate.
Even better is to ask on our [Discord](https://qdrant.to/discord) server :)

**Comment by timvisee** [comment]
We've released [Qdrant 1.12.5](https://github.com/qdrant/qdrant/releases/tag/v1.12.5) which includes the fix for this problem.
